### PR TITLE
add whitelist logic

### PR DESCRIPTION
**Pull request by JimmyShi22**

### CODE REVIEW

File: libp2p/Service.cpp [hunk 1]
```diff
@@ -94,6 +95,9 @@ void Service::heartBeat()
     {
         return;
     }
+
+    checkWhitelistAndClearSession();
```
Comment (JimmyShi22) [inline libp2p/Service.cpp]: Delete illegal sessions each heatbeat

File: libp2p/Service.cpp [hunk 2]
```diff
@@ -150,6 +160,42 @@ void Service::heartBeat()
...
 }
 
+// reset whitelist: stop session which is not in whitelist
+void Service::setWhitelist(PeerWhitelist::Ptr _whitelist)
+{
+    m_whitelist = _whitelist;
+    host()->setWhitelist(_whitelist);
+    checkWhitelistAndClearSession();
+    SERVICE_LOG(DEBUG) << LOG_BADGE("Whitelist") << LOG_DESC("Set whitelist")
+                       << m_whitelist->dump(true);
```
Comment (JimmyShi22) [inline libp2p/Service.cpp]: Delete illegal sessions once whitelist set

File: libp2p/Service.cpp [hunk 3]
```diff
@@ -154,6 +164,46 @@ void Service::heartBeat()
...
+void Service::checkWhitelistAndClearSession()
+{
+    if (m_whitelist != nullptr && m_whitelist->enable())
+    {
+        std::unordered_map<NodeID, P2PSession::Ptr> sessions;
+        {
+            RecursiveGuard l(x_sessions);
+            sessions = m_sessions;
+        }
+        for (auto session : sessions)
```
Comment (JimmyShi22) [inline libp2p/Service.cpp]: add lock here

File: libp2p/Service.cpp [hunk 4]
```diff
@@ -154,6 +164,46 @@ void Service::heartBeat()
...
+
+void Service::checkWhitelistAndClearSession()
+{
+    if (m_whitelist != nullptr && m_whitelist->enable())
+    {
+        std::unordered_map<NodeID, P2PSession::Ptr> sessions;
+        {
+            RecursiveGuard l(x_sessions);
+            sessions = m_sessions;
+        }
```
Comment (JimmyShi22) [inline libp2p/Service.cpp]: Add lock here

File: libp2p/Service.cpp [hunk 5]
```diff
@@ -127,6 +131,12 @@ void Service::heartBeat()
             SERVICE_LOG(DEBUG) << LOG_DESC("heartBeat ignore invalid address");
             continue;
         }
+        if (m_whitelist != nullptr && !m_whitelist->has(it.second))
+        {
+            SERVICE_LOG(TRACE) << LOG_DESC("heartBeat outside whitelist")
+                               << LOG_KV("nodeid", it.second.abridged());
```
Comment (bxq2011hust) [inline libp2p/Service.cpp]: If node is not in whitelist, should we disconnect it here?

Comment (JimmyShi22) [reply libp2p/Service.cpp]: Yes, this logic has been written in line 103
checkWhitelistAndClearSession();

**Pull request merged**